File: Density/app/src/main/java/org/cornelldti/density/density/util/FluxUtil.java
package org.cornelldti.density.density.util;

import android.content.res.Resources;

import org.cornelldti.density.density.R;

import java.text.SimpleDateFormat;
import java.util.Calendar;

public class FluxUtil {

    public static String getDayString() {
        String dayString = "";
        Calendar calendar = Calendar.getInstance();
        int day = calendar.get(Calendar.DAY_OF_WEEK);
        switch (day) {
            case Calendar.SUNDAY:
                dayString = "SUN";
                break;
            case Calendar.MONDAY:
                dayString = "MON";
                break;
            case Calendar.TUESDAY:
                dayString = "TUE";
                break;
            case Calendar.WEDNESDAY:
                dayString = "WED";
                break;
            case Calendar.THURSDAY:
                dayString = "THU";
                break;
            case Calendar.FRIDAY:
                dayString = "FRI";
                break;
            case Calendar.SATURDAY:
                dayString = "SAT";
                break;
        }
        return dayString;
    }

    public static String dayFullString(String day) {
        switch (day) {
            case "MON":
                return "Monday";
            case "TUE":
                return "Tuesday";
            case "WED":
                return "Wednesday";
            case "THU":
                return "Thursday";
            case "FRI":
                return "Friday";
            case "SAT":
                return "Saturday";
            case "SUN":
                return "Sunday";
        }
        return "";
    }

    // returns integer corresponding to day where 0 = Sunday, 1 = Monday, ... , 6 = Saturday
    public static int intOfDay(String day) {
        switch (day) {
            case "SUN":
                return 0;
            case "MON":
                return 1;
            case "TUE":
                return 2;
            case "WED":
                return 3;
            case "THU":
                return 4;
            case "FRI":
                return 5;
            case "SAT":
                return 6;
        }
        return -1;
    }

    public static String dateOfDay(String day, String pattern) {
        Calendar current = Calendar.getInstance();
        SimpleDateFormat format = new SimpleDateFormat(pattern);
        SimpleDateFormat checkFormat = new SimpleDateFormat("E");

        String dayCheck = checkFormat.format(current.getTime()).toUpperCase();
        while (!dayCheck.equals(day)) {
            current.add(Calendar.DAY_OF_MONTH, 1);
            dayCheck = checkFormat.format(current.getTime()).toUpperCase();
        }

        return format.format(current.getTime());
    }
}

File: Density/app/src/main/java/org/cornelldti/density/density/util/FluxUtil.kt
package org.cornelldti.density.density.util

import android.text.format.DateFormat
import org.cornelldti.density.density.DensityApplication
import java.text.SimpleDateFormat
import java.util.*

object FluxUtil {

    val daysList = listOf("SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT")

    val dayString: String
        get() = when (Calendar.getInstance().get(Calendar.DAY_OF_WEEK)) {
            Calendar.SUNDAY -> "SUN"
            Calendar.MONDAY -> "MON"
            Calendar.TUESDAY -> "TUE"
            Calendar.WEDNESDAY -> "WED"
            Calendar.THURSDAY -> "THU"
            Calendar.FRIDAY -> "FRI"
            Calendar.SATURDAY -> "SAT"
            else -> ""
        }

    val dayInt: Int
        get() = when (dayString) {
            "SUN" -> 0
            "MON" -> 1
            "TUE" -> 2
            "WED" -> 3
            "THU" -> 4
            "FRI" -> 5
            "SAT" -> 6
            else -> -1
        }

    fun dayFullString(day: String): String = when (day) {
        "MON" -> "Monday"
        "TUE" -> "Tuesday"
        "WED" -> "Wednesday"
        "THU" -> "Thursday"
        "FRI" -> "Friday"
        "SAT" -> "Saturday"
        "SUN" -> "Sunday"
        else -> ""
    }

    /**
     * This function provides the date in object format for historical endpoint request.
     */
    fun getDateObject(day: String): Date {
        val current = Calendar.getInstance()
        val checkFormat = SimpleDateFormat("E", Locale.US)

        var dayCheck = checkFormat.format(current.time).toUpperCase(Locale.US)
        while (dayCheck != day) {
            current.add(Calendar.DAY_OF_MONTH, 1)
            dayCheck = checkFormat.format(current.time).toUpperCase(Locale.US)
        }
        return current.time
    }

    fun getCurrentDateObject(): Date {
        val current = Calendar.getInstance()
        return current.time
    }


    /**
     * This function provides the date in string format for historical and operating hours
     * endpoint request.
     */
    fun convertDateObjectToString(date: Date): String {
        val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        return format.format(date)
    }

    fun getDayDifference(currentDay: String, tappedDay: String): Int {
        var x: Int = daysList.indexOf(currentDay)
        var count = 0
        while (daysList[x % 7] != tappedDay) {
            count += 1
            x += 1
        }
        return count
    }

    /**
     * @param daysAfter days after today to fetch date string for
     * @return Date String for the date that is [daysAfter] days after the current date
     */
    fun getDateStringDaysAfter(daysAfter: Int): String {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.DAY_OF_YEAR, daysAfter)
        var format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        return format.format(calendar.time)
    }

    fun getDateDaysAfter(daysAfter: Int): Date {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.DAY_OF_YEAR, daysAfter)
        return calendar.time
    }

    /**
     * This function provides the current date for the menu endpoint request.
     */
    fun getCurrentDate(): String {
        val current = Calendar.getInstance()
        var format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        return format.format(current.time)
    }

    fun parseTime(timestamp: Long): String {
        val timeZone = Calendar.getInstance().timeZone
        var format = SimpleDateFormat("h:mma", Locale.US)
        if (DateFormat.is24HourFormat(DensityApplication.getAppContext())) {
            format = SimpleDateFormat("HH:mm", Locale.US)
        }
        format.timeZone = timeZone

        return format.format(Date(timestamp * 1000)).toLowerCase(Locale.US)
    }
}
